web/text_layer_builder.js

```diff
@@ -56,6 +56,14 @@ class TextLayerBuilder {
     this.textLayerRenderTask = null;
     this.enhanceTextSelection = enhanceTextSelection;
 
+    // Listen for events about updating find matches for this page.
+    this.eventBus.on('updatetextlayermatches', (evt) => {
```

Comment:
I've just realized that this PR, and my recent find highlight one, both suffer from the same *very serious* bug. The `EventBus` listeners that are bound in this way are never actually being removed, casing the `EventBus._listeners['updatetextlayermatches'] = [...]` array to grow without bounds as new `TextLayerBuilder` instances are created :-(

Simply put, just because the `this.textLayer` property on `PDFPageView` is [cleared here](https://github.com/mozilla/pdf.js/blob/54d6c2436c17b6d0efc6fb8e36f36358eba1a9e6/web/pdf_page_view.js#L272) that does *not* imply that the `TextLayerBuilder` instance actually goes away unfortunately.

Since it's less than fifteen minutes since I found this problem, I'm still not sure if there exists a *generally* viable solution here (i.e. one not requiring ugly hacks). Also, it's now becoming clear to me why the current implementation (i.e. the one in `master`) looks the way it does, given these recently discovered problems.

Reply:
Why does the `TextLayerBuilder` instance not get deleted when it's cleared? Can't we use `this.eventBus.off` in https://github.com/mozilla/pdf.js/blob/40a6d175555316f2dbafc5bc240e52114a3e1eb7/web/text_layer_builder.js#L125 to remove the listener once rendering is being cancelled, or would that not be enough?

Reply:
I did considered that, but quite frankly it's really not at all a good idea since cancelling the `renderTextLayer` task itself doesn't necessarily mean that you'll also want to subsequently "destroy" the `TextLayerBuilder` instance too; in particular consider what would happen with multiple `TextLayerBuilder.render` calls with such a change: https://github.com/mozilla/pdf.js/blob/54d6c2436c17b6d0efc6fb8e36f36358eba1a9e6/web/text_layer_builder.js#L87-L91

The best idea I'm able to come up with on short notice, is the *first* commit that's now included in PR #10100. Please note that I'm not saying that it's a great solution, but it seems to work reasonably well and having an event fire when rendering is cancelled could be generally useful too (as a complement to 'pagerendered').